File: gmail_informes.py
import base64
from pathlib import Path

from app_paths import get_base_dir, get_data_dir, get_resource_path


SCOPES = ["https://www.googleapis.com/auth/gmail.readonly"]
TIPOS_ACEPTADOS = {
    "application/pdf",
    "application/msword",
    "application/vnd.openxmlformats-officedocument.wordprocessingml.document",
    "image/jpeg",
    "image/png",
    "image/bmp",
    "image/tiff",
}


def get_gmail_credentials_path() -> Path:
    candidates = (
        get_resource_path("resources/credentials.json"),
        get_base_dir() / "credentials" / "credentials.json",
    )
    for candidate in candidates:
        if candidate.exists():
            return candidate
    return candidates[0]


def get_gmail_token_path() -> Path:
    return get_data_dir() / "token_gmail.json"


def _load_google_deps():
    try:
        from google.auth.transport.requests import Request
        from google.oauth2.credentials import Credentials
        from google_auth_oauthlib.flow import InstalledAppFlow
        from googleapiclient.discovery import build
    except ImportError as exc:
        raise RuntimeError(
            "Faltan dependencias de Gmail. Instala requirements.txt y volve a abrir la app."
        ) from exc
    return Request, Credentials, InstalledAppFlow, build


def autenticar(
    credentials_path: Path | str | None = None,
    token_path: Path | str | None = None,
    interactive: bool = True,
):
    Request, Credentials, InstalledAppFlow, _build = _load_google_deps()
    credentials_path = Path(credentials_path or get_gmail_credentials_path())
    token_path = Path(token_path or get_gmail_token_path())

    if not credentials_path.exists():
        raise RuntimeError(f"No se encontro credentials.json de Gmail en {credentials_path}")

    creds = None
    if token_path.exists():
        creds = Credentials.from_authorized_user_file(str(token_path), SCOPES)
    if not creds or not creds.valid:
        if creds and creds.expired and creds.refresh_token:
            creds.refresh(Request())
        elif interactive:
            flow = InstalledAppFlow.from_client_secrets_file(str(credentials_path), SCOPES)
            creds = flow.run_local_server(port=0)
        else:
            raise RuntimeError("No hay token Gmail valido.")
        token_path.parent.mkdir(parents=True, exist_ok=True)
        token_path.write_text(creds.to_json(), encoding="utf-8")
    return creds


def get_connected_email(
    credentials_path: Path | str | None = None,
    token_path: Path | str | None = None,
    interactive: bool = True,
) -> str:
    _Request, _Credentials, _InstalledAppFlow, build = _load_google_deps()
    creds = autenticar(credentials_path=credentials_path, token_path=token_path, interactive=interactive)
    service = build("gmail", "v1", credentials=creds)
    profile = service.users().getProfile(userId="me").execute()
    return str(profile.get("emailAddress", "")).strip()


def descargar_informes(
    credentials_path: Path | str | None,
    token_path: Path | str | None,
    destino: Path | str,
    fecha_desde: str,
    fecha_hasta: str,
    callback_progreso=None,
    should_stop=None,
) -> int:
    """
    fecha_desde y fecha_hasta van en formato YYYY/MM/DD.
    Gmail interpreta before como fecha exclusiva.
    """
    _Request, _Credentials, _InstalledAppFlow, build = _load_google_deps()
    creds = autenticar(credentials_path=credentials_path, token_path=token_path)
    service = build("gmail", "v1", credentials=creds)

    q = f"has:attachment after:{fecha_desde} before:{fecha_hasta}"
    threads: list[dict] = []
    token = None
    while True:
        if should_stop and should_stop():
            break
        response = (
            service.users()
            .threads()
            .list(userId="me", q=q, maxResults=500, pageToken=token)
            .execute()
        )
        threads.extend(response.get("threads", []))
        token = response.get("nextPageToken")
        if not token:
            break

    destino = Path(destino)
    destino.mkdir(parents=True, exist_ok=True)
    total = 0

    for thread in threads:
        if should_stop and should_stop():
            break
        try:
            thread_data = (
                service.users()
                .threads()
                .get(userId="me", id=thread["id"], format="full")
                .execute()
            )
            for msg in thread_data.get("messages", []):
                if should_stop and should_stop():
                    break
                for part in _iter_parts(msg.get("payload", {})):
                    if should_stop and should_stop():
                        break
                    filename = str(part.get("filename", "")).strip()
                    if not filename or part.get("mimeType") not in TIPOS_ACEPTADOS:
                        continue
                    safe_filename = Path(filename).name
                    dest_file = destino / safe_filename
                    if dest_file.exists():
                        continue
                    body = part.get("body", {})
                    attachment_id = body.get("attachmentId")
                    data = body.get("data")
                    if attachment_id and not data:
                        data = (
                            service.users()
                            .messages()
                            .attachments()
                            .get(userId="me", messageId=msg["id"], id=attachment_id)
                            .execute()
                            .get("data", "")
                        )
                    if not data:
                        continue
                    dest_file.write_bytes(base64.urlsafe_b64decode(data + "=="))
                    total += 1
                    if callback_progreso:
                        callback_progreso(safe_filename)
        except Exception:
            continue

    return total


def _iter_parts(payload: dict):
    for part in payload.get("parts", []) or []:
        yield part
        yield from _iter_parts(part)

File: web/gmail_informes.js
"use strict";
// Descarga de informes desde el mail (informesnssalud@gmail.com) en Node.
// Porta la lógica de desktop-app/gmail_informes.py: busca mails con adjuntos en un
// rango de fechas y baja los PDF/Word/imagen. Usa el refresh_token ya autorizado.
const fs = require("fs");
const path = require("path");
const crypto = require("crypto");

const TIPOS_ACEPTADOS = new Set([
  "application/pdf",
  "application/msword",
  "application/vnd.openxmlformats-officedocument.wordprocessingml.document",
  "image/jpeg", "image/png", "image/bmp", "image/tiff",
]);

// Busca el token en: variable de entorno, volumen de datos, o el repo (desktop-app).
// Devuelve el objeto del token o null si no hay.
function cargarToken(dataDir) {
  const env = process.env.GMAIL_INFORMES_TOKEN;
  if (env) { try { return JSON.parse(env); } catch { /* sigue */ } }
  const candidatos = [
    dataDir && path.join(dataDir, "token_gmail_informes.json"),
    path.join(__dirname, "..", "desktop-app", "token_gmail_informes.json"),
  ].filter(Boolean);
  for (const c of candidatos) {
    try { if (fs.existsSync(c)) return JSON.parse(fs.readFileSync(c, "utf8")); } catch { /* sigue */ }
  }
  return null;
}

function clienteGmail(token) {
  const { google } = require("googleapis");
  const o = new google.auth.OAuth2(token.client_id, token.client_secret);
  o.setCredentials({ refresh_token: token.refresh_token });
  return google.gmail({ version: "v1", auth: o });
}

// Mail de la casilla conectada (para mostrar en la UI).
async function emailConectado(token) {
  const gmail = clienteGmail(token);
  const r = await gmail.users.getProfile({ userId: "me" });
  return (r.data && r.data.emailAddress) || "";
}

function* iterarPartes(payload) {
  for (const part of (payload && payload.parts) || []) {
    yield part;
    yield* iterarPartes(part);
  }
}

// Baja los adjuntos aceptados en [desde, hasta) (formato YYYY/MM/DD, before exclusivo).
// Saltea lo que ya está en la cabina. `filtros` es:
//   yaPorNombre(filename, size) -> true  (barato: NO baja el adjunto)
//   yaPorHash(sha256)           -> true  (seguro: ya lo bajó, compara contenido)
//
// Son dos pasadas a propósito. Comparar solo por NOMBRE pierde archivos:
// dos adjuntos distintos que se llaman igual ("informe.pdf", de dos
// pacientes) hacían que el segundo no entrara nunca, y eso no se nota —
// no aparece un duplicado, falta un informe. Comparar solo por CONTENIDO
// obliga a bajar todos los adjuntos del rango en cada corrida.
//
// Entonces: el nombre corta gratis cuando ADEMÁS coincide el tamaño (el
// caso normal, volver a pedir un mes ya traído), y el hash decide en el
// resto. Un mismo archivo reenviado con otro nombre tampoco entra dos
// veces, que antes sí pasaba.
//
// Devuelve [{ filename, buffer, hash, fecha, asunto }].
async function descargarAdjuntos(token, desde, hasta, filtros) {
  const yaPorNombre = (filtros && filtros.yaPorNombre) || (() => false);
  const yaPorHash = (filtros && filtros.yaPorHash) || (() => false);
  const gmail = clienteGmail(token);
  const q = `has:attachment after:${desde} before:${hasta}`;
  // Listar todos los mensajes del rango (paginado).
  const ids = [];
  let pageToken = undefined;
  do {
    const r = await gmail.users.messages.list({ userId: "me", q, maxResults: 500, pageToken });
    for (const m of (r.data.messages || [])) ids.push(m.id);
    pageToken = r.data.nextPageToken;
  } while (pageToken);

  const bajados = [];
  const vistos = new Set();
  for (const id of ids) {
    let msg;
    try { msg = await gmail.users.messages.get({ userId: "me", id, format: "full" }); }
    catch { continue; }
    // Fecha del mail (YYYY-MM-DD) para poder filtrar los informes por período.
    let fecha = "";
    try { const d = new Date(Number(msg.data.internalDate)); if (!isNaN(d)) fecha = d.toISOString().slice(0, 10); } catch { /* sin fecha */ }
    // Asunto del mail (para mostrarlo como columna en la cabina).
    let asunto = "";
    try {
      const hs = (msg.data.payload && msg.data.payload.headers) || [];
      const h = hs.find((x) => String(x.name || "").toLowerCase() === "subject");
      asunto = h ? String(h.value || "").trim().slice(0, 300) : "";
    } catch { /* sin asunto */ }
    for (const part of iterarPartes(msg.data.payload || {})) {
      const filename = String(part.filename || "").trim();
      if (!filename || !TIPOS_ACEPTADOS.has(part.mimeType)) continue;
      const safe = path.basename(filename);
      const tam = Number((part.body && part.body.size) || 0);
      // Corte barato: mismo nombre Y mismo tamaño es, con toda
      // probabilidad, el mismo archivo. Se salta sin bajarlo.
      if (yaPorNombre(safe, tam)) continue;
      let data = part.body && part.body.data;
      const attId = part.body && part.body.attachmentId;
      if (attId && !data) {
        try {
          const a = await gmail.users.messages.attachments.get({ userId: "me", messageId: id, id: attId });
          data = a.data && a.data.data;
        } catch { continue; }
      }
      if (!data) continue;
      const buffer = Buffer.from(data, "base64");
      const hash = crypto.createHash("sha256").update(buffer).digest("hex");
      // Ahora sí, por contenido: cubre el mismo archivo con otro nombre
      // y el mismo mail reenviado.
      if (vistos.has(hash)) continue;   // repetido dentro de esta corrida
      if (yaPorHash(hash)) continue;    // ya está en la cabina
      vistos.add(hash);
      bajados.push({ filename: safe, buffer, hash, fecha, asunto });
    }
  }
  return bajados;
}

module.exports = { cargarToken, emailConectado, descargarAdjuntos, TIPOS_ACEPTADOS };
